pandas/core/dtypes/base.py

```diff
@@ -236,6 +236,10 @@ def construct_from_string(cls, string: str):
         """
         if not isinstance(string, str):
             raise TypeError(f"Expects a string, got {type(string).__name__}")
+
+        # error: Non-overlapping equality check (left operand type: "str", right
+        #  operand type: "Callable[[ExtensionDtype], str]")  [comparison-overlap]
```

Comment:
can this use the self._ensure_type ?

Reply:
i don't think that's applicable here.

the assert is to avoid an ignore on the following line (the comparison).

the comparison would still fail strict_equality even if all subclasses override this method/override name attribute. This is because the type checker is checking this method using the name attribute in the base class. i.e. checking the base class isolation.

I think this is related to "Python language's lack of notation for abstract class attributes" see https://github.com/python/mypy/issues/7760

so without the assert an ignore would be required here.

However, the assert prevents creating subclasses that call this method when not using a class attribute for name. 

This could instead be enforced by mypy instead of the assert and i believe the implementation of name in the base class should be...
```diff
diff --git a/pandas/core/dtypes/base.py b/pandas/core/dtypes/base.py
index 1b4e7062b..5235aac91 100644
--- a/pandas/core/dtypes/base.py
+++ b/pandas/core/dtypes/base.py
@@ -1,6 +1,6 @@
 """Extend pandas with custom array types"""
 from typing import Any, List, Optional, Tuple, Type
-
+import abc
 import numpy as np
 
 from pandas.errors import AbstractMethodError
@@ -8,7 +8,7 @@ from pandas.errors import AbstractMethodError
 from pandas.core.dtypes.generic import ABCDataFrame, ABCIndexClass, ABCSeries
 
 
-class ExtensionDtype:
+class ExtensionDtype(metaclass=abc.ABCMeta):
     """
     A custom data type, to be paired with an ExtensionArray.
 
@@ -162,13 +162,15 @@ class ExtensionDtype:
         return "O"
 
     @property
-    def name(self) -> str:
+    @classmethod
+    @abc.abstractmethod
+    def name(cls) -> str:
         """
         A string identifying the data type.
 
         Will be used for display in, e.g. ``Series.dtype``
         """
-        raise AbstractMethodError(self)
+        pass
 
     @property
     def names(self) -> Optional[List[str]]:
``` 

However, mypy would give the following errors...

pandas\core\dtypes\base.py:241: error: Non-overlapping equality check (left operand type: "str", right operand type: "Callable[[], str]")
pandas\core\dtypes\dtypes.py:754: error: Signature of "name" incompatible with supertype "ExtensionDtype"
pandas\core\dtypes\dtypes.py:895: error: Signature of "name" incompatible with supertype "ExtensionDtype"
pandas\tests\extension\test_common.py:44: error: Cannot instantiate abstract class 'DummyDtype' with abstract attribute 'name'

the first is due to https://github.com/python/mypy/issues/7760, the next two are because PeriodDType.name and DatetimeTZDtype.name are not implemented using class attributes